### PR TITLE
Multiple chooser panel 

### LINKED ISSUE #2203: InlinePanel Bulk Chooser
This repeats the call in #1717, except I don't see a reason this should be limited to ImageChooserBlocks. I'm also not calling for changes to the panel UI, which would be more of an undertaking, rather a streamlining of what we already have.

One pattern we regularly repeat in our sites is to wrap a ChooserPanel, with an InlinePanel.

Sometimes it's a SnippetChooserPanel, pointing to a "Category" snippet, in effect replicating free tagging except where there is essentially less freedom to make mistakes. Or perhaps it's an ImageChooserPanel, used to generate an image gallery.

Using an InlinePanel in these situations is quite clumsy:
1. Click the plus at the top of the InlinePanel and the panel opens. 
2. Click the "Choose thing" button, the modal opens, you choose the desired item.
3. To add the next, click the "Add thing" button
4. Click "Choose thing" _again_, again the modal opens, again the item is chosen
   ... repeat

In both scenarios, the user probably already knows _all_ the things they plan to add, before they even get to step 1. Even if they don't, they certainly shouldn't need to open 10 modals one after the other, to choose 10 things.

What would make more sense, is an interface where instead of a blank Panel appearing when you click in step 1, you are instead immediately presented with a modal in which you can choose multiple items at once, by adding a checkmark to every item in a list. Paginating through results in the modal would remember your item choices in previous pages. Having finally chosen everything, the modal would close and a panel for each of the items would be generated at once.

### PR DESCRIPTION
**Pull request by gasman**
Fixes #2203

Provides a new panel type `MultipleChooserPanel` that extends `InlinePanel` so that the "Add" button opens the chooser modal, allowing multiple items to be selected and resulting in the appropriate number of child panels being inserted and populated. This follows from the prototype built as https://github.com/wagtail/wagtail-multiple-chooser-panel/.

The comment thread starting at https://github.com/wagtail/wagtail/issues/2203#issuecomment-1330552555 will help a lot in following what's going on here. There's a lot of preliminary refactoring of the chooser widget and modal JS so that ChooserFactory objects (the client-side equivalent of the BaseChooser widget object) can be defined alongside the Chooser object (which represents a single rendered instance of the chooser widget) and share bits of implementation with minimal duplicated code. This also brings PageChooserFactory onto the same base ChooserFactory class. All of this allows us to add an `openModal` endpoint to ChooserFactory so that we can open the chooser modal without needing an actual instance of the chooser.

To assist in testing, I've created this branch of bakerydemo with several MultipleChooserPanels added to BlogPage - https://github.com/gasman/bakerydemo/tree/core-multiple-chooser-panel

### CODE REVIEW

File: wagtail/admin/views/generic/chooser.py [hunk 1]
```diff
@@ -145,11 +197,21 @@ def columns(self):
                 "title",
                 label=_("Title"),
                 accessor=str,
-                url_name=self.chosen_url_name,
+                get_url=(
```
Comment (laymonage) [inline wagtail/admin/views/generic/chooser.py:200]: Should we remove the URL if it's a multiple chooser? If we still have the URL, accidentally clicking on the title after selecting many of the checkboxes will close the modal with only selecting the one clicked.

**Review by laymonage — COMMENTED** [review]
I'm very happy with this, it works really well. Thank you!

I just have some minor comments and a concern about keeping the individual items' links.

File: client/src/components/MultipleChooserPanel/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            const formPrefix = `${opts.formsetPrefix}-${formIndex}`;
+            const chooserFieldId = `${formPrefix}-${opts.chooserFieldName}`;
+            const chooserWidget =
+              this.chooserWidgetFactory.getById(chooserFieldId);
+            chooserWidget.setStateFromModalData(item);
+          });
+        },
+        { multiple: true },
+      );
+    });
```
[the remark this answers is not in the document]
Comment (gasman) [reply client/src/components/MultipleChooserPanel/index.js]: Good point! Updated in f2111a5a84b15f4da99534ed9891a487eada5c6c - there will always be exactly one element with the `-OPEN_MODAL` ID, so I've simplified this a bit with a plain `getElementById`.

File: client/src/entrypoints/documents/document-chooser.js [hunk 3]
```diff
@@ -1,11 +1,18 @@
...
+  chooserModalClass = DocumentChooserModal;
 }
 window.DocumentChooser = DocumentChooser;
 
+class DocumentChooserFactory extends ChooserFactory {
+  widgetClass = DocumentChooser;
+  // eslint-disable-next-line no-undef
+  chooserModalClass = DocumentChooserModal;
+}
+window.DocumentChooserFactory = DocumentChooserFactory;
```
[the remark this answers is not in the document]
Comment (gasman) [reply client/src/entrypoints/documents/document-chooser.js:14]: Was in two minds about this - it does mean we end up loading rather a lot of copies of the base ChooserWidget code, even if that's only ~1.5Kb a time. But thinking about how much faff it was to get the dependency ordering right in the Python-side media declarations on the widget / adapter objects, it's probably worth it to get rid of some of that global variable spaghetti :-) Will update.

[the remark this answers is not in the document]
Comment (gasman) [reply client/src/entrypoints/documents/document-chooser.js]: Changed to use imports (and along the way, moved their definitions to client/src/components) in 2609e4d2cfef0c9a273ec62da11840a73e07a360.

File: wagtail/admin/views/generic/chooser.py (hunk 1 above)
Comment (gasman) [reply wagtail/admin/views/generic/chooser.py:215]: As discussed, have now changed the title into a label for the checkbox in 9621d35b02fda8b659e8220f72e613841a3ffaab.

**Review by laymonage — COMMENTED** [review]
Loving the change to use `<label>`, it's much easier to use now! As I noted in the stand-up earlier, currently the multiple chooser doesn't seem to take the `max_num` parameter into account. Not sure how hard it is to validate that in the chooser, but at the very least I think we should disable the "Add" button (that launches the modal) when it reaches the `max_num`, just like the current chooser does.

P.s. another great addition would be if the chooser can somehow keep track of the order of the checked items, so that eventually the order is also preserved when the items get added to the form's list. But I think is just a nice-to-have, considering the added complexity of having to separately keep track of the selected items instead of relying on the form.

**Comment by gasman** [comment]
@laymonage `max_num` now enforced in a5abd28ceb8975d18423a747a3b2d284f4cdad18.

**Review by laymonage — APPROVED** [review]
Can confirm that `max_num` is now enforced with how I'd expect it to work. Awesome work, thanks!

**Comment by gasman** [comment]
Merged in d7adcec58432602eebddead019305f2febbd4871. Thanks @laymonage!

**Pull request closed without merging**